src/components/CheckboxInput/CheckboxInput.tsx

```diff
@@ -25,76 +24,69 @@ export type CheckboxInputProps = CheckboxHTMLElementProps & {
   id: string;
 };
 
-const checkboxIconNameMap = {
-  indeterminate: 'checkbox-indeterminate',
-  true: 'checkbox-checked',
-  false: 'checkbox-unchecked',
-};
-const CheckboxSvg = ({
-  checked,
-  disabled,
-}: {
-  checked: boolean | 'indeterminate';
-  disabled?: boolean;
-}) => {
-  const iconClassName = clsx(
-    styles['checkbox__icon'],
-    disabled && styles['checkbox__icon--disabled'],
-  );
-  return (
-    <Icon
-      className={iconClassName}
-      name={
-        checkboxIconNameMap[`${checked}`] as
-          | 'checkbox-indeterminate'
-          | 'checkbox-checked'
-          | 'checkbox-unchecked'
-      }
-      purpose="decorative"
-      size="1.5rem"
-    />
-  );
-};
-
 /**
  * Checkbox input element, exported for greater flexibility.
  * You must provide an `id` prop and connect it to a visible label.
  */
 export const CheckboxInput = React.forwardRef<
   HTMLInputElement,
   CheckboxInputProps
->(({ checked = false, className, disabled, ...other }, ref) => {
+>(({ checked, className, disabled, ...other }, ref) => {
   // Make indeterminate checkbox visually match the colors of a
   // checked state, but announce itself as "mixed" to screen readers
   const checkedProps =
     checked === 'indeterminate'
       ? {
           'aria-checked': 'mixed' as const,
-          checked: true,
+          // Indeterminate checkboxes are unchecked when submitted to the server.
+          checked: false,
         }
       : {
           checked,
         };
 
+  const safeRef = useForwardedRef(ref);
+
+  // If the checkbox is "indeterminate", set that property on the node. This can only be done in JS
+  // for some reason.
+  // See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#indeterminate_state_checkboxes
+  React.useEffect(() => {
+    if (safeRef.current) {
+      safeRef.current.indeterminate = checked === 'indeterminate';
+    }
+  }, [checked, safeRef]);
+
   return (
-    <span
-      className={clsx(
-        styles['input__wrapper'],
-        disabled && styles['input__wrapper--disabled'],
-      )}
-    >
-      <input
-        className={clsx(className, styles['checkbox__input'])}
-        data-bootstrap-override="checkbox"
-        disabled={disabled}
-        ref={ref}
-        type="checkbox"
-        {...checkedProps}
-        {...other}
-      />
-      <CheckboxSvg checked={checked} disabled={disabled} />
-    </span>
+    <input
+      className={clsx(className, styles['checkbox__input'])}
+      data-bootstrap-override="checkbox"
+      disabled={disabled}
+      ref={safeRef}
+      type="checkbox"
+      {...checkedProps}
+      {...other}
+    />
   );
 });
 
 CheckboxInput.displayName = 'CheckboxInput';
+
+/**
+ * Use a forwarded ref if present, or create one if not.
+ */
+function useForwardedRef<T>(ref: React.ForwardedRef<T>): React.RefObject<T> {
```

Comment:
TODO: extract this hook to another file, if we end up going with this.

Idea is to use a forwarded ref, which may or may not be provided.

An alternative I've seen (which is basically the same as this, just different semantics) is to come up with a scheme to combine refs:

```ts
const myRef = useRef();
const combinedRef = useCombinedRefs(forwardedRef, myRef);
```